File: 1/1190/ReverseSubstringsBetweenEachPairOfParentheses.java
/*
https://leetcode.com/problems/reverse-substrings-between-each-pair-of-parentheses/description/?envType=daily-question&envId=2024-07-11
*/

class Solution {
    public String reverseParentheses(String s) {
        StringBuilder temporaryS=new StringBuilder(s);
        while(true){
            int beginIndexForFirstClosingBracket=0, indexForLastOpenBracket=0;
            StringBuilder resultantReversedInnermostString=new StringBuilder();
            boolean isAnyClosingBracketAvailable=false;
            while(true){
                if(beginIndexForFirstClosingBracket>=temporaryS.length()) break;
                if(temporaryS.charAt(beginIndexForFirstClosingBracket)==')'){
                    isAnyClosingBracketAvailable=true;
                    temporaryS.setCharAt(beginIndexForFirstClosingBracket, '#');
                    beginIndexForFirstClosingBracket--;
                    indexForLastOpenBracket=beginIndexForFirstClosingBracket;
                    boolean isAnyOpeningBracketAvailable=true;
                    while(true){
                        if(temporaryS.charAt(indexForLastOpenBracket)=='(') break;
                        indexForLastOpenBracket--;
                        if(indexForLastOpenBracket<0){
                            isAnyOpeningBracketAvailable=false;
                            break;
                        }
                    }
                    if(isAnyOpeningBracketAvailable){
                        temporaryS.setCharAt(indexForLastOpenBracket, '#');
                        indexForLastOpenBracket++;
                        for(int i=beginIndexForFirstClosingBracket; i>=indexForLastOpenBracket; --i){
                            resultantReversedInnermostString.append(temporaryS.charAt(i));
                        }
                        int beginIdxForReversedString=0;
                        for(int i=indexForLastOpenBracket; i<=beginIndexForFirstClosingBracket; ++i){
                            temporaryS.setCharAt(i, resultantReversedInnermostString.charAt(beginIdxForReversedString++));
                        }
                        break;
                    } else break;
                }
                beginIndexForFirstClosingBracket++;
            }
            if(!isAnyClosingBracketAvailable) break;
            for(int i=0; i<temporaryS.length(); ++i){
                if(temporaryS.charAt(i)=='#'){
                    temporaryS.deleteCharAt(i);
                    --i;
                }
            }
        }
        return String.valueOf(temporaryS);
    }
}

File: 1/1000s/11s/1190/ReverseSubstringsBetweenEachPairOfParentheses.cpp
/*
https://leetcode.com/problems/reverse-substrings-between-each-pair-of-parentheses/description/?envType=daily-question&envId=2024-07-11
*/

class Solution {
public:
    string reverseParentheses(string s) {
        int sz=s.size();
        while(true){
            int idx1=0, idx2=0;
            string tofrm="";
            bool flag=false;
            while(true){
                if(s[idx1]==')'){
                    flag=true;
                    s[idx1]='#';
                    idx1--;
                    idx2=idx1;
                    bool secflag=false;
                    while(true){
                        if(s[idx2]=='(') break;
                        idx2--;
                        if(idx2<0){
                            secflag=true;
                            break;
                        }
                    }
                    if(!secflag){
                        s[idx2]='#';
                        idx2++;
                        for(int i=idx1; i>=idx2; --i){
                            tofrm+=s[i];
                        }
                        int nwidx=0;
                        for(int i=idx2; i<=idx1; ++i){
                            s[i]=tofrm[nwidx++];
                        }
                        break;
                    } else break;
                }
                idx1++;
                if(idx1>=sz) break;
            }
            if(!flag) break;
            s.erase(remove(s.begin(), s.end(), '#'), s.end());
        }
        return s;
    }
};